### ISSUE TITLE
cannot differentiate `log`

### ISSUE THREAD

**Issue by scheidan** [wsmoses/Enzyme.jl#163]
I realize that differentiating through `log` fails:
```Julia
using Enzyme

foo(x) = log(x)

autodiff(foo, Active(2.0)) # ERROR: LLVM error: unknown binary operator, see below for details
```

<details>
  <summary>complete error message </summary>

```Julia
define internal fastcc double @preprocess_julia__log_7692(double %0) unnamed_addr #0 !dbg !471 {
top:
  %1 = call {}*** @julia.get_pgcstack()
  %2 = fcmp ule double %0, 0.000000e+00, !dbg !472
  br i1 %2, label %L113, label %L3, !dbg !474

L3:                                               ; preds = %top
  %3 = fcmp une double %0, 0x7FF0000000000000, !dbg !475
  br i1 %3, label %L7, label %L6, !dbg !476

L6:                                               ; preds = %L116, %L113, %L3
  %merge = phi double [ 0x7FF0000000000000, %L3 ], [ 0xFFF0000000000000, %L113 ], [ 0x7FF8000000000000, %L116 ]
  ret double %merge, !dbg !476

L7:                                               ; preds = %L3
  %4 = fcmp ule double %0, 0x3FEE0FABFBC702A3, !dbg !477
  %5 = fcmp uge double %0, 0x3FF1082B577D34EE, !dbg !478
  %6 = or i1 %4, %5, !dbg !478
  br i1 %6, label %L45, label %L14, !dbg !478

L14:                                              ; preds = %L7
  %7 = fadd double %0, -1.000000e+00, !dbg !479
  %8 = fadd double %7, 2.000000e+00, !dbg !481
  %9 = fdiv double 1.000000e+00, %8, !dbg !484
  %10 = fmul double %7, 2.000000e+00, !dbg !485
  %11 = fmul double %10, %9, !dbg !485
  %12 = fmul double %11, %11, !dbg !488
  %13 = fmul contract double %12, 0x3F3C8034C85DFFF0, !dbg !490
  %14 = fadd contract double %13, 0x3F62492307F1519F, !dbg !490
  %15 = fmul contract double %12, %14, !dbg !490
  %16 = fadd contract double %15, 0x3F89999999BAC6D4, !dbg !490
  %17 = fmul contract double %12, %16, !dbg !490
  %18 = fadd contract double %17, 0x3FB55555555554E6, !dbg !490
  %19 = fmul double %11, %12, !dbg !494
  %20 = fmul double %19, %18, !dbg !494
  %21 = bitcast double %11 to i64, !dbg !496
  %22 = and i64 %21, -134217728, !dbg !499
  %23 = bitcast i64 %22 to double, !dbg !496
  %24 = bitcast double %7 to i64, !dbg !500
  %25 = and i64 %24, -134217728, !dbg !503
  %26 = bitcast i64 %25 to double, !dbg !500
  %27 = fsub double %7, %26, !dbg !504
  %28 = fsub double %7, %23, !dbg !506
  %29 = fmul double %28, 2.000000e+00, !dbg !508
  %30 = fmul double %26, %23, !dbg !508
  %31 = fsub double %29, %30, !dbg !506
  %32 = fmul double %27, %23, !dbg !508
  %33 = fsub double %31, %32, !dbg !506
  %34 = fmul double %9, %33, !dbg !508
  %35 = fadd double %34, %20, !dbg !509
  %36 = fmul double %23, 0.000000e+00, !dbg !511
  %37 = fadd double %36, %35, !dbg !512
  %38 = fadd double %37, %23, !dbg !512
  ret double %38, !dbg !483

L45:                                              ; preds = %L7
  %39 = bitcast double %0 to i64, !dbg !514
  %40 = lshr i64 %39, 52, !dbg !516
  %41 = and i64 %40, 2047, !dbg !519
  %.not = icmp eq i64 %41, 0, !dbg !521
  br i1 %.not, label %L62, label %L79, !dbg !522

L62:                                              ; preds = %L45
  %42 = fmul double %0, 0x4350000000000000, !dbg !523
  %43 = bitcast double %42 to i64, !dbg !525
  %44 = lshr i64 %43, 52, !dbg !527
  %45 = and i64 %44, 2047, !dbg !530
  %46 = add nsw i64 %45, -54, !dbg !532
  br label %L79, !dbg !532

L79:                                              ; preds = %L62, %L45
  %value_phi2 = phi i64 [ %46, %L62 ], [ %41, %L45 ]
  %value_phi3 = phi i64 [ %43, %L62 ], [ %39, %L45 ]
  %47 = add nsw i64 %value_phi2, -1023, !dbg !533
  %48 = and i64 %value_phi3, 4503599627370495, !dbg !535
  %49 = or i64 %48, 4607182418800017408, !dbg !537
  %50 = bitcast i64 %49 to double, !dbg !538
  %51 = sitofp i64 %47 to double, !dbg !539
  %52 = fadd double %50, 0x42C0000000000000, !dbg !541
  %53 = fadd double %52, 0xC2C0000000000000, !dbg !543
  %54 = fsub double %50, %53, !dbg !544
  %55 = fmul double %53, 1.280000e+02, !dbg !546
  %56 = fptosi double %55 to i64, !dbg !548
  %57 = freeze i64 %56, !dbg !548
  %58 = add i64 %57, -128, !dbg !549
  %59 = getelementptr inbounds [129 x [2 x double]], [129 x [2 x double]]* @_j_const1, i64 0, i64 %58, i64 0, !dbg !552
  %60 = getelementptr inbounds [129 x [2 x double]], [129 x [2 x double]]* @_j_const1, i64 0, i64 %58, i64 1, !dbg !553
  %61 = fmul double %51, 0x3FE62E42FEFA4000, !dbg !554
  %62 = load double, double* %59, align 16, !dbg !556, !tbaa !169
  %63 = fadd double %62, %61, !dbg !556
  %64 = fmul double %51, 0x3D48432A1B0E2634, !dbg !557
  %65 = load double, double* %60, align 8, !dbg !559, !tbaa !169
  %66 = fsub double %65, %64, !dbg !559
  %67 = fmul double %54, 2.000000e+00, !dbg !560
  %68 = fadd double %53, %50, !dbg !562
  %69 = fdiv double %67, %68, !dbg !563
  %70 = fmul double %69, %69, !dbg !564
  %71 = fmul contract double %70, 0x3F8999A0BC712416, !dbg !566
  %72 = fadd contract double %71, 0x3FB5555555550286, !dbg !566
  %73 = fmul double %69, %70, !dbg !570
  %74 = fmul double %73, %72, !dbg !570
  %75 = fadd double %66, %74, !dbg !572
  %76 = fadd double %69, %75, !dbg !572
  %77 = fmul double %63, 0.000000e+00, !dbg !574
  %78 = fadd double %77, %76, !dbg !575
  %79 = fadd double %63, %78, !dbg !575
  ret double %79, !dbg !551

L113:                                             ; preds = %top
  %80 = fcmp une double %0, 0.000000e+00, !dbg !577
  br i1 %80, label %L116, label %L6, !dbg !578

L116:                                             ; preds = %L113
  %81 = fcmp ord double %0, 0.000000e+00, !dbg !579
  br i1 %81, label %L119, label %L6, !dbg !581

L119:                                             ; preds = %L116
  %82 = call fastcc nonnull {} addrspace(10)* @julia_throw_complex_domainerror_7697() #14, !dbg !582
  unreachable, !dbg !582
}

 constantarg[double %0] = 0 type: {[-1]:Float@double} - vals: {}
 constantinst[  %1 = call {}*** @julia.get_pgcstack()] = 1 val:1 type: {}
 constantinst[  %2 = fcmp ule double %0, 0.000000e+00, !dbg !51] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br i1 %2, label %L113, label %L3, !dbg !57] = 1 val:1 type: {}
 constantinst[  %3 = fcmp une double %0, 0x7FF0000000000000, !dbg !58] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br i1 %3, label %L7, label %L6, !dbg !60] = 1 val:1 type: {}
 constantinst[  %merge = phi double [ 0x7FF0000000000000, %L3 ], [ 0xFFF0000000000000, %L113 ], [ 0x7FF8000000000000, %L116 ]] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  ret double %merge, !dbg !60] = 1 val:1 type: {}
 constantinst[  %4 = fcmp ule double %0, 0x3FEE0FABFBC702A3, !dbg !61] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %5 = fcmp uge double %0, 0x3FF1082B577D34EE, !dbg !62] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %6 = or i1 %4, %5, !dbg !62] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br i1 %6, label %L45, label %L14, !dbg !62] = 1 val:1 type: {}
 constantinst[  %7 = fadd double %0, -1.000000e+00, !dbg !63] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %8 = fadd double %7, 2.000000e+00, !dbg !66] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %9 = fdiv double 1.000000e+00, %8, !dbg !71] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %10 = fmul double %7, 2.000000e+00, !dbg !73] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %11 = fmul double %10, %9, !dbg !73] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %12 = fmul double %11, %11, !dbg !78] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %13 = fmul contract double %12, 0x3F3C8034C85DFFF0, !dbg !80] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %14 = fadd contract double %13, 0x3F62492307F1519F, !dbg !80] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %15 = fmul contract double %12, %14, !dbg !80] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %16 = fadd contract double %15, 0x3F89999999BAC6D4, !dbg !80] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %17 = fmul contract double %12, %16, !dbg !80] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %18 = fadd contract double %17, 0x3FB55555555554E6, !dbg !80] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %19 = fmul double %11, %12, !dbg !87] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %20 = fmul double %19, %18, !dbg !87] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %21 = bitcast double %11 to i64, !dbg !89] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %22 = and i64 %21, -134217728, !dbg !95] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %23 = bitcast i64 %22 to double, !dbg !89] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %24 = bitcast double %7 to i64, !dbg !98] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %25 = and i64 %24, -134217728, !dbg !101] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %26 = bitcast i64 %25 to double, !dbg !98] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %27 = fsub double %7, %26, !dbg !102] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %28 = fsub double %7, %23, !dbg !104] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %29 = fmul double %28, 2.000000e+00, !dbg !106] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %30 = fmul double %26, %23, !dbg !106] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %31 = fsub double %29, %30, !dbg !104] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %32 = fmul double %27, %23, !dbg !106] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %33 = fsub double %31, %32, !dbg !104] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %34 = fmul double %9, %33, !dbg !106] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %35 = fadd double %34, %20, !dbg !107] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %36 = fmul double %23, 0.000000e+00, !dbg !109] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %37 = fadd double %36, %35, !dbg !110] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %38 = fadd double %37, %23, !dbg !110] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  ret double %38, !dbg !70] = 1 val:1 type: {}
 constantinst[  %39 = bitcast double %0 to i64, !dbg !115] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %40 = lshr i64 %39, 52, !dbg !117] = 1 val:1 type: {}
 constantinst[  %41 = and i64 %40, 2047, !dbg !121] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %.not = icmp eq i64 %41, 0, !dbg !123] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br i1 %.not, label %L62, label %L79, !dbg !126] = 1 val:1 type: {}
 constantinst[  %42 = fmul double %0, 0x4350000000000000, !dbg !127] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %43 = bitcast double %42 to i64, !dbg !129] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %44 = lshr i64 %43, 52, !dbg !131] = 1 val:1 type: {}
 constantinst[  %45 = and i64 %44, 2047, !dbg !134] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %46 = add nsw i64 %45, -54, !dbg !136] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br label %L79, !dbg !136] = 1 val:1 type: {}
 constantinst[  %value_phi2 = phi i64 [ %46, %L62 ], [ %41, %L45 ]] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %value_phi3 = phi i64 [ %43, %L62 ], [ %39, %L45 ]] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %47 = add nsw i64 %value_phi2, -1023, !dbg !138] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %48 = and i64 %value_phi3, 4503599627370495, !dbg !140] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %49 = or i64 %48, 4607182418800017408, !dbg !142] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %50 = bitcast i64 %49 to double, !dbg !144] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %51 = sitofp i64 %47 to double, !dbg !145] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %52 = fadd double %50, 0x42C0000000000000, !dbg !148] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %53 = fadd double %52, 0xC2C0000000000000, !dbg !150] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %54 = fsub double %50, %53, !dbg !151] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %55 = fmul double %53, 1.280000e+02, !dbg !153] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %56 = fptosi double %55 to i64, !dbg !155] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %57 = freeze i64 %56, !dbg !155] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %58 = add i64 %57, -128, !dbg !157] = 1 val:1 type: {[-1]:Integer}
 constantinst[  %59 = getelementptr inbounds [129 x [2 x double]], [129 x [2 x double]]* @_j_const1, i64 0, i64 %58, i64 0, !dbg !163] = 1 val:1 type: {[-1]:Pointer, [-1,0]:Float@double}
 constantinst[  %60 = getelementptr inbounds [129 x [2 x double]], [129 x [2 x double]]* @_j_const1, i64 0, i64 %58, i64 1, !dbg !165] = 1 val:1 type: {[-1]:Pointer, [-1,0]:Float@double}
 constantinst[  %61 = fmul double %51, 0x3FE62E42FEFA4000, !dbg !166] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %62 = load double, double* %59, align 16, !dbg !168, !tbaa !169] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %63 = fadd double %62, %61, !dbg !168] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %64 = fmul double %51, 0x3D48432A1B0E2634, !dbg !175] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %65 = load double, double* %60, align 8, !dbg !177, !tbaa !169] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %66 = fsub double %65, %64, !dbg !177] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %67 = fmul double %54, 2.000000e+00, !dbg !178] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %68 = fadd double %53, %50, !dbg !180] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %69 = fdiv double %67, %68, !dbg !181] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %70 = fmul double %69, %69, !dbg !182] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %71 = fmul contract double %70, 0x3F8999A0BC712416, !dbg !184] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %72 = fadd contract double %71, 0x3FB5555555550286, !dbg !184] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %73 = fmul double %69, %70, !dbg !188] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %74 = fmul double %73, %72, !dbg !188] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %75 = fadd double %66, %74, !dbg !190] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %76 = fadd double %69, %75, !dbg !190] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %77 = fmul double %63, 0.000000e+00, !dbg !192] = 1 val:1 type: {[-1]:Float@double}
 constantinst[  %78 = fadd double %77, %76, !dbg !193] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %79 = fadd double %63, %78, !dbg !193] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  ret double %79, !dbg !162] = 1 val:1 type: {}
 constantinst[  %80 = fcmp une double %0, 0.000000e+00, !dbg !195] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br i1 %80, label %L116, label %L6, !dbg !196] = 1 val:1 type: {}
 constantinst[  %81 = fcmp ord double %0, 0.000000e+00, !dbg !197] = 1 val:1 type: {[-1]:Integer}
 constantinst[  br i1 %81, label %L119, label %L6, !dbg !201] = 1 val:1 type: {}
 constantinst[  %82 = call fastcc nonnull {} addrspace(10)* @julia_throw_complex_domainerror_7697() #14, !dbg !202] = 1 val:1 type: {}
 constantinst[  unreachable, !dbg !202] = 1 val:1 type: {}
cannot handle unknown binary operator:   %48 = and i64 %value_phi3, 4503599627370495, !dbg !140
ERROR: LLVM error: unknown binary operator
Stacktrace:
  [1] handle_error(reason::Cstring)
    @ LLVM ~/.julia/packages/LLVM/shuW4/src/core/context.jl:105
  [2] EnzymeCreatePrimalAndGradient(logic::Enzyme.Logic, todiff::LLVM.Function, retType::
Enzyme.API.CDIFFE_TYPE, constant_args::Vector{Enzyme.API.CDIFFE_TYPE}, TA::Enzyme.TypeAnalysis, returnValue::Bool, dretUsed::Bool, topLevel::Bool, additionalArg::Ptr{Nothing}, typeInfo::Enzyme.FnTypeInfo, uncacheable_args::Vector{Bool}, augmented::Ptr{Nothing}, atomicAdd::Bool, postOpt::Bool)
    @ Enzyme.API ~/.julia/packages/Enzyme/cAQ4a/src/api.jl:108
  [3] enzyme!(job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}}, mod::LLVM.Module, primalf::LLVM.Function, adjoint::GPUCompiler.FunctionSpec{typeof(foo), Tuple{Active{Float64}}}, split::Bool, parallel::Bool, actualRetType::Type, dupClosure::Bool)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/cAQ4a/src/compiler.jl:1740
  [4] codegen(output::Symbol, job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/cAQ4a/src/compiler.jl:2282
  [5] _thunk(job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(foo), Tuple{Float64}}})
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/cAQ4a/src/compiler.jl:2623
  [6] cached_compilation(cache::Dict{UInt64, Any}, job::GPUCompiler.CompilerJob, compiler
::typeof(Enzyme.Compiler._thunk), linker::typeof(Enzyme.Compiler._link))                
    @ GPUCompiler ~/.julia/packages/GPUCompiler/HeCT6/src/cache.jl:90
  [7] thunk(f::typeof(foo), df::Nothing, ::Type{Active{Float64}}, tt::Type{Tuple{Active{Float64}}}, ::Val{false})
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/cAQ4a/src/compiler.jl:2675
  [8] autodiff
    @ ~/.julia/packages/Enzyme/cAQ4a/src/Enzyme.jl:186 [inlined]
  [9] autodiff(f::typeof(foo), args::Active{Float64})
    @ Enzyme ~/.julia/packages/Enzyme/cAQ4a/src/Enzyme.jl:216
 [10] top-level scope
    @ REPL[158]:1
```
</details>

I'm using Julia 1.7 and Enzyme v0.8.0 on Linux. All tests of Enzyme pass.

**Comment by vchuravy** [comment]
Yeah the problem is that the `log` function got inlined and now we are trying to differentiate through bit-twiddling.

**Comment by vchuravy** [comment]
Should be fixed by #164
